src/fault/mutex/shared_mutex.cpp

```diff
@@ -0,0 +1,106 @@
+#include <yaclib/fault/detail/mutex/shared_mutex.hpp>
+
+// TODO(myannyax) avoid copypaste from shared_timed_mutex
+namespace yaclib::detail {
+
+void SharedMutex::lock() {
+  auto me = yaclib_std::this_thread::get_id();
+  assert(_exclusive_owner != me);
+  {
+    std::shared_lock lock(_helper_m);
+    assert(_shared_owners.find(me) == _shared_owners.end());
+  }
+
+  InjectFault();
+  _m.lock();
+  InjectFault();
+
+  _exclusive_owner = me;
+}
+
+bool SharedMutex::try_lock() noexcept {
+  auto me = yaclib_std::this_thread::get_id();
+  assert(_exclusive_owner != me);
+  {
+    std::shared_lock lock(_helper_m);
+    assert(_shared_owners.find(me) == _shared_owners.end());
+  }
+  InjectFault();
+  auto res = _m.try_lock();
+  InjectFault();
+
+  if (res) {
+    _exclusive_owner = yaclib_std::this_thread::get_id();
+  }
+  return res;
+}
+
+void SharedMutex::unlock() noexcept {
+  assert(_exclusive_owner != yaclib::detail::kInvalidThreadId);
+  assert(_exclusive_owner == yaclib_std::this_thread::get_id());
+
+  _exclusive_owner = yaclib::detail::kInvalidThreadId;
+
+  InjectFault();
+  _m.unlock();
+  InjectFault();
+}
+
+void SharedMutex::lock_shared() {
+  auto me = yaclib_std::this_thread::get_id();
+  assert(_exclusive_owner != me);
+  {
+    std::shared_lock lock(_helper_m);
+    assert(_shared_owners.find(me) == _shared_owners.end());
+  }
+
+  InjectFault();
+  _m.lock_shared();
+  InjectFault();
+
+  {
+    std::unique_lock lock(_helper_m);
+    _shared_owners.insert(me);
+  }
+}
+
+bool SharedMutex::try_lock_shared() noexcept {
+  auto me = yaclib_std::this_thread::get_id();
+  assert(_exclusive_owner != me);
+  {
+    std::shared_lock lock(_helper_m);
+    assert(_shared_owners.find(me) == _shared_owners.end());
+  }
+
+  InjectFault();
+  auto res = _m.try_lock_shared();
+  InjectFault();
+
+  if (res) {
+    {
+      std::unique_lock lock(_helper_m);
+      _shared_owners.insert(me);
+    }
+  }
+  return res;
+}
+
+void SharedMutex::unlock_shared() noexcept {
+  auto me = yaclib_std::this_thread::get_id();
+  {
+    std::shared_lock lock(_helper_m);
+    assert(_shared_owners.find(me) != _shared_owners.end());
+  }
+
+  {
+    std::unique_lock lock(_helper_m);
+    _shared_owners.erase(me);
+  }
+  _exclusive_owner = yaclib::detail::kInvalidThreadId;
+
+  InjectFault();
+  _m.unlock_shared();
+  InjectFault();
```

Comment:
Please add a macro for this pattern, this will protect us from errors of misuse and also allow us to transfer information about the source to asserts.